cdap-watchdog/src/main/java/io/cdap/cdap/metrics/collect/AggregatedMetricsEmitter.java

```diff
@@ -33,38 +34,64 @@
   private final String name;
   // metric value
   private final AtomicLong value;
-  // specifies if the metric type is gauge or counter
-  private final AtomicBoolean gaugeUsed;
+
+  private volatile MetricType metricType = MetricType.COUNTER;
+  private Distribution distribution;
 
   AggregatedMetricsEmitter(String name) {
     if (name == null || name.isEmpty()) {
       LOG.warn("Creating emmitter with " + (name == null ? "null" : "empty") + " name, ");
     }
 
     this.name = name;
-    this.value = new AtomicLong();
-    this.gaugeUsed = new AtomicBoolean(false);
+    this.value = new AtomicLong(0);
   }
 
   void increment(long value) {
     this.value.addAndGet(value);
+    this.metricType = MetricType.COUNTER;
   }
 
 
   @Override
   public MetricValue emit() {
-    // todo CDAP-2195 - potential race condition , reseting value and type has to be done together
-    long value = this.value.getAndSet(0);
-    MetricType type = gaugeUsed.getAndSet(false) ? MetricType.GAUGE : MetricType.COUNTER;
-    return new MetricValue(name, type, value);
+    if (metricType == MetricType.DISTRIBUTION) {
+      // TODO it is possible the distribution has all 0s. Skip such MetricValues
+      Distribution oldVal;
+      synchronized (this) {
+        oldVal = distribution;
+        distribution = new Distribution();
+      }
+      if (oldVal != null) {
+        LOG.debug(String.format("Emitting distribution metric: %s", oldVal.toString()));
+        return oldVal.getMetricValue(name);
+      } else {
+        return new Distribution().getMetricValue(name);
+      }
+    } else {
```

Comment:
No need to have `else` since the `if` block always returns.

Reply:
Done